design-proposals/descheduler-support.md

```diff
@@ -0,0 +1,106 @@
+
+| Migration Strategy    | VMI is migrateable                    | VMI is not migrateable | Remarks                               |
+|-----------------------|---------------------------------------|------------------------|---------------------------------------|
+| None                  | N/A                                   | N/A                    | Always approve                        |
+| LiveMigrate           | Mark the VMI for eviction and approve | Deny                   |                                       |
+| LiveMigrateIfPossible | Mark the VMI for eviction and approve | Approve                |                                       |
+| External              | N/A                                   | N/A                    | Mark the VMI for eviction and approve |
+
+
+### Disruption Budget Controller
+`virt-launcher` pods have another layer of protection against evictions using [PDBs](https://kubernetes.io/docs/concepts/workloads/pods/disruptions/#pod-disruption-budgets).
+
+The controller is responsible for managing PDBs:
+
+The controller decides whether a VMI needs protection against eviction based on its migration strategy (defined on the VMI with a fallback to a cluster-wide definition):
+
+| Migration Strategy    | Requires a PDB                 | 
+|-----------------------|--------------------------------|
+| None                  | False                          |
+| LiveMigrate           | True                           |
+| LiveMigrateIfPossible | Only if the VMI is migrateable |
+| External              | True                           |
+
+### Evacuation Controller
+The controller looks for potential VMIs to evict and tries to migrate them to another node.
+
+### Migration Controller
+The controller manages the migration process initiated by the Evacuation controller.
+The controller migrates the `virt-launcher` pod from the source node to a target node.
+After the migration process is completed, the original `virt-launcher` pod is deleted from the source node and the eviction process ends.
+
+## Motivation
+Cluster administrators want their workloads to be balanced across the cluster.
+These workloads include KubeVirt VMs.
+
+## Goals
+1. Enable KubeVirt to work on existing K8s clusters that already have a descheduler deployed.
+2. Enable cluster admins to introduce a descheduler to an existing cluster with KubeVirt deployed.
+
+## Non Goals
+The descheduler should respect the two annotations included in this document.
+
+## Definition of Users
+- Cluster administrator
+
+## User Stories
+- As a cluster admin running a K8s cluster with a descheduler, I want to deploy and use KubeVirt.
+- As a cluster admin running a K8s cluster with KubeVirt, I want to deploy a descheduler to load balance all my workloads.
+- As a cluster admin I want to enable cluster-wide descheduler support.
+
+## Repos
+1. kubevirt/kubevirt
+2. kubevirt/hyperconverged-cluster-operator
+
+# Design
```

Comment:
Would you please mention the current and expected descheduler workflow? Something between the following lines:
Current workflow:
1. iterate through pods (the descheduling framework performs some filtering to reduce the set of pods to process)
2. sort the pods (the descheduling framework allows to implemented custom sorters as well)
3. iterate through filtered and sorted pods and perform eviction until either all such pods are evicted or a limit is reached (e.g. pods per node, pods per namespace). If the eviction is successful, increase counter of evicted pods. If not, move to another pod.

Expected workflow: extend the current workflow with:
- before evicting the pods sort them based on `descheduler.alpha.kubernetes.io/request-evict-only` and `descheduler.alpha.kubernetes.io/eviction-in-progress` annotation to prefer annotated pods first
- if `descheduler.alpha.kubernetes.io/eviction-in-progress` annotation is present, do not invoke the eviction API, only increase the pod counter
- if `descheduler.alpha.kubernetes.io/request-evict-only` annotation is present, invoke the eviction API.

Now about the eviction API error:
- 200 return code should not happen for annotated pods given "request-evict-only" is expected
- 500 return code means something went wrong => do not increase the pod counter
- 429 return code means eviction is scheduled => increase the pod counter.

Reply:
Nevertheless, given the validating webhook is s custom implementation and there's no upstream eviction API equivalent for "eviction request confirmed" return code the descheduler will need to keep the feature as alpha until there's an upstream/community accepted solution for it. Temporarily, the descheduler can either make the conclusion on the request based on the custom error code or parsing the error message text.

Reply:
I believe these steps should go into the descheduler's requirement document which is the counterpart to this documet.